File: src/data/colors.ts
import type { Color } from './types';

export const WHEEL_ORDER: Color[] = [
  'Red', 'Vermilion', 'Orange', 'Amber', 'Yellow', 'Chartreuse',
  'Green', 'Teal', 'Blue', 'Indigo', 'Purple', 'Magenta'
];

export const ALL_COLORS = WHEEL_ORDER;

export function colorIndex(color: Color): number {
  return WHEEL_ORDER.indexOf(color);
}

const PRIMARY_INDICES = new Set([0, 4, 8]);    // Red, Yellow, Blue
const SECONDARY_INDICES = new Set([2, 6, 10]); // Orange, Green, Purple

// Can two colors be mixed?
// Valid mixes: primary+primary, or primary+secondary that are 2 apart on the wheel.
export function canMix(a: Color, b: Color): boolean {
  const ai = colorIndex(a);
  const bi = colorIndex(b);
  const diff = Math.abs(ai - bi);
  const distance = Math.min(diff, WHEEL_ORDER.length - diff);

  if (PRIMARY_INDICES.has(ai) && PRIMARY_INDICES.has(bi)) {
    return ai !== bi;
  }

  const oneIsPrimary = PRIMARY_INDICES.has(ai) || PRIMARY_INDICES.has(bi);
  const oneIsSecondary = SECONDARY_INDICES.has(ai) || SECONDARY_INDICES.has(bi);
  return oneIsPrimary && oneIsSecondary && distance === 2;
}

// Returns the color produced by mixing a and b.
// Throws if they can't be mixed.
export function mixResult(a: Color, b: Color): Color {
  if (!canMix(a, b)) throw new Error(`Cannot mix ${a} and ${b}`);
  const ai = colorIndex(a);
  const bi = colorIndex(b);
  const n = WHEEL_ORDER.length;
  // Find midpoint along the shorter arc
  const forwardDist = (bi - ai + n) % n;
  if (forwardDist <= n / 2) {
    return WHEEL_ORDER[(ai + forwardDist / 2) % n];
  } else {
    const backDist = n - forwardDist;
    return WHEEL_ORDER[(ai - backDist / 2 + n) % n];
  }
}

// CSS color for display
export function colorToHex(color: Color): string {
  const map: Record<Color, string> = {
    'Red': '#e63946',
    'Vermilion': '#e76f51',
    'Orange': '#f4a261',
    'Amber': '#e9c46a',
    'Yellow': '#f2e205',
    'Chartreuse': '#a7c957',
    'Green': '#2d6a4f',
    'Teal': '#219ebc',
    'Blue': '#264653',
    'Indigo': '#3a0ca3',
    'Purple': '#7209b7',
    'Magenta': '#d63384',
  };
  return map[color];
}

export function blendColors(colors: Color[]): string {
  if (colors.length === 0) return '#000000';
  const hexes = colors.map(c => colorToHex(c));
  let r = 0, g = 0, b = 0;
  for (const hex of hexes) {
    r += parseInt(hex.slice(1, 3), 16);
    g += parseInt(hex.slice(3, 5), 16);
    b += parseInt(hex.slice(5, 7), 16);
  }
  const n = hexes.length;
  r = Math.round(r / n);
  g = Math.round(g / n);
  b = Math.round(b / n);
  return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
}

export function textColorForBackground(hex: string): string {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  const yiq = (r * 299 + g * 587 + b * 114) / 1000;
  return yiq >= 128 ? '#000000' : '#ffffff';
}

File: colori-core/src/colors.rs
use crate::types::{Color, ColorWheel, NUM_COLORS};

fn is_primary(idx: usize) -> bool {
    idx == 0 || idx == 4 || idx == 8
}

fn is_secondary(idx: usize) -> bool {
    idx == 2 || idx == 6 || idx == 10
}

pub fn can_mix(a: Color, b: Color) -> bool {
    let a_index = a.index();
    let b_index = b.index();
    let diff = if a_index > b_index { a_index - b_index } else { b_index - a_index };
    let distance = diff.min(NUM_COLORS - diff);

    if is_primary(a_index) && is_primary(b_index) {
        return a_index != b_index;
    }

    let one_is_primary = is_primary(a_index) || is_primary(b_index);
    let one_is_secondary = is_secondary(a_index) || is_secondary(b_index);
    one_is_primary && one_is_secondary && distance == 2
}

pub fn mix_result(a: Color, b: Color) -> Color {
    let a_index = a.index();
    let b_index = b.index();
    let n = NUM_COLORS;
    let forward_dist = (b_index + n - a_index) % n;
    if forward_dist <= n / 2 {
        Color::from_index((a_index + forward_dist / 2) % n)
    } else {
        let back_dist = n - forward_dist;
        Color::from_index((a_index + n - back_dist / 2) % n)
    }
}

pub const PRIMARIES: [Color; 3] = [Color::Red, Color::Yellow, Color::Blue];
pub const SECONDARIES: [Color; 3] = [Color::Orange, Color::Green, Color::Purple];
pub const TERTIARIES: [Color; 6] = [
    Color::Vermilion,
    Color::Amber,
    Color::Chartreuse,
    Color::Teal,
    Color::Indigo,
    Color::Magenta,
];

/// The 9 color pairs that can legally be mixed:
/// 3 primary+primary and 6 primary+adjacent_secondary.
pub const VALID_MIX_PAIRS: [(Color, Color); 9] = [
    // Primary + Primary
    (Color::Red, Color::Yellow),
    (Color::Red, Color::Blue),
    (Color::Yellow, Color::Blue),
    // Primary + Adjacent Secondary
    (Color::Red, Color::Orange),
    (Color::Yellow, Color::Orange),
    (Color::Yellow, Color::Green),
    (Color::Blue, Color::Green),
    (Color::Blue, Color::Purple),
    (Color::Red, Color::Purple),
];

#[inline]
pub fn perform_mix(wheel: &mut ColorWheel, a: Color, b: Color) -> bool {
    if !can_mix(a, b) {
        return false;
    }
    if wheel.get(a) == 0 || wheel.get(b) == 0 {
        return false;
    }
    wheel.decrement(a);
    wheel.decrement(b);
    let result = mix_result(a, b);
    wheel.increment(result);
    true
}

/// Performs a mix without checking `can_mix` or wheel amounts.
/// The caller must guarantee that `a` and `b` are a valid mix pair
/// and that the wheel has at least one of each.
#[inline]
pub fn perform_mix_unchecked(wheel: &mut ColorWheel, a: Color, b: Color) {
    wheel.decrement(a);
    wheel.decrement(b);
    let result = mix_result(a, b);
    wheel.increment(result);
}

#[inline]
pub fn can_pay_cost(wheel: &ColorWheel, cost: &[Color]) -> bool {
    let mut used = [0u32; 12];
    for &c in cost {
        let idx = c.index();
        let needed = used[idx] + 1;
        if wheel.get(c) < needed {
            return false;
        }
        used[idx] = needed;
    }
    true
}

#[inline]
pub fn pay_cost(wheel: &mut ColorWheel, cost: &[Color]) -> bool {
    if !can_pay_cost(wheel, cost) {
        return false;
    }
    for &c in cost {
        wheel.decrement(c);
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_primary_mix() {
        assert!(can_mix(Color::Red, Color::Yellow));
        assert!(can_mix(Color::Red, Color::Blue));
        assert!(can_mix(Color::Yellow, Color::Blue));
        assert!(!can_mix(Color::Red, Color::Red));
    }

    #[test]
    fn test_primary_mix_results() {
        assert_eq!(mix_result(Color::Red, Color::Yellow), Color::Orange);
        assert_eq!(mix_result(Color::Yellow, Color::Blue), Color::Green);
        assert_eq!(mix_result(Color::Red, Color::Blue), Color::Purple);
    }

    #[test]
    fn test_secondary_tertiary_mix() {
        // Red + Orange = Vermilion
        assert!(can_mix(Color::Red, Color::Orange));
        assert_eq!(mix_result(Color::Red, Color::Orange), Color::Vermilion);

        // Yellow + Orange = Amber
        assert!(can_mix(Color::Yellow, Color::Orange));
        assert_eq!(mix_result(Color::Yellow, Color::Orange), Color::Amber);
    }

    #[test]
    fn test_cannot_mix_secondaries() {
        assert!(!can_mix(Color::Orange, Color::Green));
        assert!(!can_mix(Color::Orange, Color::Purple));
        assert!(!can_mix(Color::Green, Color::Purple));
    }

    #[test]
    fn test_cannot_mix_non_adjacent() {
        assert!(!can_mix(Color::Red, Color::Green));
        assert!(!can_mix(Color::Yellow, Color::Purple));
    }

    #[test]
    fn test_valid_mix_pairs_is_exhaustive() {
        use crate::types::ALL_COLORS;

        // Every pair in VALID_MIX_PAIRS must pass can_mix
        for &(a, b) in &VALID_MIX_PAIRS {
            assert!(can_mix(a, b), "{:?} and {:?} should be mixable", a, b);
        }

        // Every mixable pair must appear in VALID_MIX_PAIRS
        for i in 0..ALL_COLORS.len() {
            for j in (i + 1)..ALL_COLORS.len() {
                let a = ALL_COLORS[i];
                let b = ALL_COLORS[j];
                if can_mix(a, b) {
                    let found = VALID_MIX_PAIRS.iter().any(|&(x, y)| {
                        (x == a && y == b) || (x == b && y == a)
                    });
                    assert!(found, "Mixable pair ({:?}, {:?}) missing from VALID_MIX_PAIRS", a, b);
                }
            }
        }
    }
}
